app/src/main/java/org/stepik/android/presentation/catalog/CatalogPresenter.kt

```diff
@@ -0,0 +1,116 @@
+import org.stepic.droid.di.qualifiers.MainScheduler
+import org.stepic.droid.features.stories.presentation.StoriesPresenter
+import org.stepic.droid.model.StepikFilter
+import org.stepic.droid.preferences.SharedPreferenceHelper
+import org.stepic.droid.util.emptyOnErrorStub
+import org.stepik.android.domain.catalog.interactor.CatalogInteractor
+import org.stepik.android.domain.course_list.model.CourseListQuery
+import org.stepik.android.presentation.course_list.CourseListCollectionPresenter
+import org.stepik.android.presentation.course_list.CourseListQueryPresenter
+import org.stepik.android.view.injection.catalog.FiltersBus
+import ru.nobird.android.presentation.base.DisposableViewModel
+import ru.nobird.android.presentation.base.PresenterBase
+import java.util.EnumSet
+import javax.inject.Inject
+import javax.inject.Provider
+
+class CatalogPresenter
+@Inject
+constructor(
+    @FiltersBus
+    private val filtersObservable: Observable<EnumSet<StepikFilter>>,
+    @BackgroundScheduler
+    private val backgroundScheduler: Scheduler,
+    @MainScheduler
+    private val mainScheduler: Scheduler,
+    private val catalogInteractor: CatalogInteractor,
+    private val storiesPresenter: StoriesPresenter,
+    private val tagsPresenter: TagsPresenter,
+    private val filtersPresenter: FiltersPresenter,
+    private val courseListCollectionPresenterProvider: Provider<CourseListCollectionPresenter>,
+    private val courseListQueryPresenterProvider: Provider<CourseListQueryPresenter>,
+    private val sharedPreferenceHelper: SharedPreferenceHelper
+) : PresenterBase<CatalogView>() {
+
+    companion object {
+        val POPULAR_COURSES_QUERY = CourseListQuery(
+            page = 1,
+            order = CourseListQuery.Order.ACTIVITY_DESC,
+            isExcludeEnded = true,
+            isPublic = true
+        )
+    }
+
+    private var state: CatalogView.State = CatalogView.State.Idle
+        set(value) {
+            field = value
+            view?.setState(value)
+        }
+
+    override val nestedDisposables: List<DisposableViewModel>
+        get() = (state as? CatalogView.State.Content)
+            ?.collections
+            ?.filterIsInstance<DisposableViewModel>()
+            ?: emptyList()
+
+    init {
+        subscribeForFilterUpdates()
+    }
+
+    fun fetchCollections(forceUpdate: Boolean = false) {
+        if (state != CatalogView.State.Idle && !forceUpdate) return
+        compositeDisposable += catalogInteractor
+            .fetchCourseCollections()
+            .subscribeOn(backgroundScheduler)
+            .observeOn(mainScheduler)
+            .subscribeBy(
+                onSuccess = { courseCollections ->
+                    val base = listOf<CatalogItem>(storiesPresenter, tagsPresenter, filtersPresenter)
+                    val collections = courseCollections.map {
+                        courseListCollectionPresenterProvider.get().apply { setDataToPresenter(it) }
```

Comment:
it seems that you have to detach & dispose previous nested presenters (from old state)